File: opencl/02_matrix_mult/main.py
import numpy as np
from cll import CL
import pyopencl as cl

class myAlgo(CL):
    def __init__(self):
        CL.__init__(self)

    def runAlgo(self):
        """
        The program implementation
        """
        #initialize client side (CPU) arrays
        N = 100
        A_VAL = .5
        B_VAL = 1
        size = N * N
        h_A = np.empty(size).astype(np.float32)
        h_B = np.empty(size).astype(np.float32)
        h_A.fill(A_VAL)
        h_B.fill(B_VAL)

        #create OpenCL buffers
        d_A = self.vectToBuffer(h_A)
        d_B = self.vectToBuffer(h_B)
        d_C = self.outBuffer(h_A.nbytes)

        np.set_printoptions(threshold='nan')
        # execute program
        mmul = self.program.mmul
        mmul.set_scalar_arg_dtypes([np.int32, None, None, None])
        mmul(self.queue, (N, N), None,
             N, d_A, d_B, d_C)
        print "First problem solved"

        h_C = np.empty_like(h_A)
        self.bufferToVect(d_C, h_C)
        print "{}".format(h_C)

        localmem = cl.LocalMemory(np.dtype(np.float32).itemsize * N)
        n_blocks = 10
        mmul2 = self.program.mmul2
        mmul2.set_scalar_arg_dtypes([np.int32, None, None, None, None])
        mmul2(self.queue, (N,), (N/n_blocks,),
              N, d_A, d_B, d_C, localmem)
        print "Second problem solved"

        h_C = np.empty_like(h_A)
        self.bufferToVect(d_C, h_C)
        print "{}".format(h_C)

        blocksize = 10
        A_block = cl.LocalMemory(np.dtype(np.float32).itemsize*blocksize**2)
        B_block = cl.LocalMemory(np.dtype(np.float32).itemsize*blocksize**2)
        mmul3 = self.program.mmul3
        mmul3.set_scalar_arg_dtypes([np.int32, None, None, None, None, None])
        mmul3(self.queue, (N, N), (blocksize, blocksize),
              N, d_A, d_B, d_C, A_block, B_block)

        print "Third problem solved"


        h_C = np.empty_like(h_A)
        self.bufferToVect(d_C, h_C)
        print "{}".format(h_C)

        # print result
        # print "a", h_A
        # print "b", h_B
        # print "c", h_C

if __name__ == "__main__":
    example = myAlgo()
    example.loadProgram("mmul.cl")
    example.runAlgo()

File: opencl/02_matrix_mult/main.cpp
#include "cll.hpp"

class myAlgo: public CL
{
public:
    // function to use for vector add
    cl::make_kernel<int, cl::Buffer, cl::Buffer, cl::Buffer> mmul;
    cl::make_kernel<int, cl::Buffer, cl::Buffer, cl::Buffer, cl::LocalSpaceArg> mmul2;
    cl::make_kernel<int, cl::Buffer, cl::Buffer, cl::Buffer, cl::LocalSpaceArg, cl::LocalSpaceArg> mmul3;
    myAlgo(std::string clfile):CL(clfile),
                               mmul(clprogram, "mmul"),
                               mmul2(clprogram, "mmul2"),
                               mmul3(clprogram, "mmul3"){
    }
    virtual void runAlgo();
    virtual ~myAlgo(){}
};

void myAlgo::runAlgo(){
    int N = 100;
    int size = N*N;
    // create vectors: 2 inputs and 1 output
    std::vector<float> h_A(size);
    std::vector<float> h_B(size);
    std::vector<float> h_C (size);

    // create buffers used as vectors in opencl
    cl::Buffer d_A;
    cl::Buffer d_B;
    cl::Buffer d_C;

    // set values of the vectors
    for(int i = 0; i < size; i++) {
        h_A[i]  = .5;
        h_B[i]  = 1;
    }

    // pass the value of the vectors to the buffers
    vectToBuffer(h_A, d_A);
    vectToBuffer(h_B, d_B);

    // create a buffer with the proper lenght for result
    d_C = outBuffer(size);

    // run opencl function
    mmul(cl::EnqueueArgs(queue, cl::NDRange(N, N)),
          N, d_A, d_B, d_C);

    // pass the result information into a vector
    bufferToVect(d_C, h_C);
    // printing
    std::cout << "First result:\n";
    for (int i = 0; i < size; i++) {
        std::cout << h_C[i] << ", ";
    }
    std::cout << "\n";

    // run opencl function
    int n_blocks(10);
    cl::LocalSpaceArg localmem = cl::Local(sizeof(float) * N);
    mmul2(cl::EnqueueArgs(queue, cl::NDRange(N),cl::NDRange(N/n_blocks)),
         N, d_A, d_B, d_C, localmem);

    // pass the result information into a vector
    bufferToVect(d_C, h_C);
    // printing
    std::cout << "Second result:\n";
    for (int i = 0; i < size; i++) {
        std::cout << h_C[i] << ", ";
    }
    std::cout << "\n";

    // run opencl function
    int blocksize(10);
    cl::LocalSpaceArg A_block = cl::Local(sizeof(float)*blocksize*blocksize);
    cl::LocalSpaceArg B_block = cl::Local(sizeof(float)*blocksize*blocksize);
    mmul3(cl::EnqueueArgs(queue, cl::NDRange(N, N),cl::NDRange(blocksize, blocksize)),
          N, d_A, d_B, d_C, A_block, B_block);

    // pass the result information into a vector
    bufferToVect(d_C, h_C);
    // printing
    std::cout << "Third result:\n";
    for (int i = 0; i < size; i++) {
        std::cout << h_C[i] << ", ";
    }
    std::cout << "\n";

}

int main(int argc, char *argv[])
{
    std::string clfile = "mmul.cl";
    myAlgo example(clfile);
    example.runAlgo();
    return 0;
}
